dwapi/files_test.go

```diff
@@ -42,12 +78,121 @@ func TestFileService_Delete(t *testing.T) {
 	}
 	endpoint := fmt.Sprintf("/datasets/%s/%s/files/%s", owner, id, filename)
 	mux.HandleFunc(endpoint, handler)
-	got, err := client.File.Delete(owner, id, filename)
+	got, err := dw.File.Delete(owner, id, filename)
 	if assert.NoError(t, err) {
 		assert.Equal(t, want, got)
 	}
 }
 
+func TestFileService_Download(t *testing.T) {
+	setup()
+	defer teardown()
+
+	want := "test content"
+
+	owner := testClientOwner
+	id := "my-awesome-dataset"
+	filename := "test-file"
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, r.Method, GET, "Expected method 'GET', got %s", r.Method)
+		fmt.Fprintf(w, `test content`)
+	}
+	endpoint := fmt.Sprintf("/file_download/%s/%s/%s", owner, id, filename)
+	mux.HandleFunc(endpoint, handler)
+	r, err := dw.File.Download(owner, id, filename)
+	if assert.NoError(t, err) {
+		got, _ := ioutil.ReadAll(r)
+		assert.Equal(t, want, string(got))
+	}
+	r.Close()
+}
+
+func TestFileService_DownloadAndSave(t *testing.T) {
+	setup()
+	defer teardown()
+
+	filename := "test-file"
+	path := filepath.Join(os.TempDir(), filename)
+	want := SuccessResponse{
+		fmt.Sprintf("File saved to %s", path),
+	}
+
+	owner := testClientOwner
+	id := "my-awesome-dataset"
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, r.Method, GET, "Expected method 'GET', got %s", r.Method)
+		fmt.Fprintf(w, `test content`)
+	}
+	endpoint := fmt.Sprintf("/file_download/%s/%s/%s", owner, id, filename)
+	mux.HandleFunc(endpoint, handler)
+	got, err := dw.File.DownloadAndSave(owner, id, filename, path)
+	if assert.NoError(t, err) {
+		assert.Equal(t, want, got)
+		assert.FileExists(t, path)
+
+		c, err := ioutil.ReadFile(path)
+		if err != nil {
+			log.Fatal(err)
+		}
+		assert.Equal(t, "test content", string(c))
+	}
+	_ = os.Remove(path)
```

Comment:
This is a tough one. I gave this approach a shot just now by converting `teardown` into a variadic function and moving the call to it in this function, to down there. While it works great, I would also argue that it makes things more inconsistent and/or slightly tougher to read.

There are a couple of `r.Close` statements, that like `os.Remove` are cleanup steps and begs the question, why aren't those in `teardown` as well? I don't feel that either should be since the file is created in this method and it feels like it should also be cleaned up here. Looking at `setup`, it could also accept a path and create a file there. As a whole, all of this seems like unnecessary complexity. My opinions here aren't terribly strong but I do prefer it as is. Do you feel that it's circumventing established best practices?

Reply:
My opinions aren't strong on this, either! Thanks for giving the suggestion a spin.